File: android/ProgrammingIdeas/Fragments/RegisterDialog.cs
using Android.App;
using Android.OS;
using Android.Support.Design.Widget;
using Android.Views;
using Android.Widget;
using Helpers;
using Newtonsoft.Json;
using ProgrammingIdeas.Api;
using ProgrammingIdeas.Helpers;
using ProgrammingIdeas.Models;
using System;
using System.Threading.Tasks;
using AlertDialog = Android.Support.V7.App.AlertDialog;
using DialogFragment = Android.Support.V4.App.DialogFragment;

namespace ProgrammingIdeas.Fragments
{
    internal class RegisterDialog : DialogFragment
    {
        private ProgressBar loadingCircle;

        public event EventHandler<string> OnUserLoggedIn;

        public override Dialog OnCreateDialog(Bundle savedInstanceState)
        {
            var view = LayoutInflater.From(App.CurrentActivity).Inflate(Resource.Layout.register_dialog, null, false);
            var emailTb = view.FindViewById<EditText>(Resource.Id.emailTb);
            var passwordTb = view.FindViewById<EditText>(Resource.Id.passwordTb);
            var retypePasswordTb = view.FindViewById<EditText>(Resource.Id.retypePasswordTb);
            var loginBtn = view.FindViewById<Button>(Resource.Id.loginBtn);
            loadingCircle = view.FindViewById<ProgressBar>(Resource.Id.loadingCircle);

            loginBtn.Click += delegate
            {
                using (var validator = new Validator())
                {
                    loadingCircle.Visibility = ViewStates.Visible;
                    validator.ValidateIsNotEmpty(emailTb, true);
                    validator.ValidateIsSame(passwordTb, retypePasswordTb);

                    if (validator.PassedValidation)
                    {
                        var data = new LoginData { Email = emailTb.Text.Trim(), Password = passwordTb.Text.Trim() };
                        RegisterUser(data);
                    }
                }
            };

            return new AlertDialog.Builder(Activity)
                .SetTitle("Sign up")
                .SetView(view)
                .Create();
        }

        private async Task RegisterUser(LoginData data)
        {
            var response = await IdeaBagApi.Instance.RegisterAsync(data);

            if (response.Payload != null)
            {
                Global.AuthUser(response.Payload);

                var expirationDate = DateTime.Now.AddMilliseconds(response.Payload.ExpiresInMillis);
                PreferenceManager.Instance.AddEntry("expiresIn", JsonConvert.SerializeObject(expirationDate));

                OnUserLoggedIn?.Invoke(this, response.Payload.Email);
                Dismiss();
            }
            else
                Snackbar.Make(loadingCircle, response.ErrorMessage, Snackbar.LengthLong).Show();

            loadingCircle.Visibility = ViewStates.Gone;
        }
    }
}

File: android/kotlin/app/src/main/java/com/alansa/ideabag2/dialogs/RegisterDialog.kt
package com.alansa.ideabag2.dialogs

import android.support.v7.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import com.alansa.ideabag2.R
import com.alansa.ideabag2.extensions.using
import com.alansa.ideabag2.utils.Validator
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.dialog_register.view.*

class RegisterDialog(private val ctx : Context, private val onRegisterSuccess: () -> Unit) : DialogFragment() {
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        var view = LayoutInflater.from(activity).inflate(R.layout.dialog_register, null)

        view.registerBtn.setOnClickListener {

            val validator = Validator()
            using(validator) {
                val email = view.emailTb.text.toString()
                val pwd = view.passwordTb.text.toString()
                validator.validateIsNotEmpty(view.emailTIL, email)
                validator.validateIsNotEmpty(view.passwordTIL, pwd)

                if (validator.passedValidation && view.passwordTb.text.toString() == view.retypePasswordTb.text.toString()) {
                    view.loadingCircle.visibility = View.VISIBLE
                    view.registerBtn.isEnabled = false

                    FirebaseAuth
                            .getInstance()
                            .createUserWithEmailAndPassword(email, pwd)
                            .addOnCompleteListener {
                                view.loadingCircle.visibility = View.GONE
                                view.registerBtn.isEnabled = true

                                Toast.makeText(ctx, if (it.isSuccessful) "Sign up successful." else it.exception?.message, Toast.LENGTH_LONG).show()

                                if (it.isSuccessful) {
                                    onRegisterSuccess()
                                    dismiss()
                                }
                            }
                } else {
                    Toast.makeText(ctx, R.string.required, Toast.LENGTH_LONG).show()
                }
            }
        }

        return AlertDialog.Builder(ctx)
                .setTitle(R.string.register)
                .setView(view)
                .create()
    }
}